float8_experimental/inference.py

```diff
@@ -0,0 +1,232 @@
+                input, self.weight.to_original_precision()
+            )
+
+        x_fp8 = cast_to_float8_e4m3fn(
+            input,
+            self.forward_config,
+            static_quantization_scale=self.static_quantization_scale,
+        )
+        return torch.nn.functional.linear(x_fp8, self.weight, self.bias)
+
+    # Builder functions for Float8LinearInference
+    def quantize_weight(self, dtype: torch.dtype = e4m3_dtype) -> None:
+        """This functions converts the weight to a Float8Tensor and sets its requires_grad to False.
+
+        Args:
+            dtype: The dtype to quantize the weight to. Default is e4m3_dtype.
+
+        Note:
+            This function is typically called during inference to quantize the weight once since
+            the weight is not updated during inference.
+
+        """
+        assert not isinstance(
+            self.weight, Float8Tensor
+        ), "Weight has already been quantized, cannot quantize again."
+        scale = tensor_to_scale(self.weight, dtype)
+        quantized_weight = to_fp8_no_autograd(
+            self.weight,
+            scale,
+            dtype,
+            self.forward_config,
+        )
+        self.weight = nn.Parameter(quantized_weight)
+        self.weight.requires_grad = False
+
+    def set_weight_and_bias(
+        self, weight: torch.nn.Parameter, bias: Optional[torch.nn.Parameter]
+    ):
+        self.weight = weight
+        self.bias = bias
+
+    def set_quantization_config(
+        self,
+        quant_config: QuantConfig,
+    ):
+        # We destructure the quant_config into the individual fields
+        # If an activation config is passed in we want to register that as a buffer
+        self.activation_casting: ActivationCasting = quant_config.activation_casting
+
+        if self.activation_casting == ActivationCasting.STATIC:
+            self.register_buffer(
+                "static_quantization_scale", quant_config.static_quantization_scale
+            )
+        else:
+            self.static_quantization_scale = None
+
+    @classmethod
+    def from_float(
+        cls, module: nn.Module, quant_config: QuantConfig, use_fast_accum: bool
+    ) -> "Float8LinearInference":
+        """
+        Create an nn.Linear with fp8 compute from another nn.Linear
+
+        Args:
+            mod (torch.nn.Linear): nn.Linear to convert
+            quant_config (QuantConfig): Configuration for the weight and activation casting
+        """
+        forward_config = ScaledMMConfig(
+            False, use_fast_accum, pad_inner_dim=config.pad_inner_dim
+        )
+        linear = cls(
+            quant_config,
+            forward_config,
+            module.in_features,
+            module.out_features,
+            False,
+            device=torch.device("meta"),
+        )
+        linear.set_weight_and_bias(module.weight, module.bias)
+        linear.quantize_weight()
+        return linear
+
+
+def cast_to_float8_e4m3fn(
```

Comment:
nit: it's a bit confusing to have this function which is the same name as the training `cast_to_float8_e4m3fn` but does a different thing, even if it lives in a different file.  Maybe either split the code into static vs dynamic or make the name clearer?

note that in https://github.com/pytorch-labs/float8_experimental/pull/290/files the training version got renamed to `cast_to_float8_e4m3_dynamic`, I'm not married to that, I just hope whatever names we use are as easy to understand as possible